sys/arduino/wireport.cpp

```diff
@@ -0,0 +1,245 @@
+ * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ * Modified 2012 by Todd Krein (todd@krein.org) to implement repeated starts
+ * ```
+ *
+ * The documentation in this file is partially extracted from the original
+ * [Arduino Reference](https://www.arduino.cc/en/Reference/Wire) of this
+ * library which is published under the
+ * [Creative Commons Attribution-ShareAlike 3.0 License]
+ * (https://creativecommons.org/licenses/by-sa/3.0/).
+ */
+
+extern "C" {
+#include <stdlib.h>
+#include <string.h>
+#include <inttypes.h>
+}
+
+#include "Wire.h"
+
+#include "log.h"
+#include "panic.h"
+#include "periph/i2c.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/* Initialize Class Variables */
+
+uint8_t WirePort::rxBuffer[WIREPORT_BUFFER_LENGTH];
+uint8_t WirePort::rxBufferIndex = 0;
+uint8_t WirePort::rxBufferLength = 0;
+
+uint8_t WirePort::txAddress = 0;
+uint8_t WirePort::txBuffer[WIREPORT_BUFFER_LENGTH];
+uint8_t WirePort::txBufferIndex = 0;
+uint8_t WirePort::txBufferLength = 0;
+
+uint8_t WirePort::transmitting = 0;
+
+/* Constructors */
+
+WirePort::WirePort(void)
+{
+    DEBUG("[wire] %s\n", __func__);
+}
+
+/* Public Methods */
+
+void WirePort::begin(void)
+{
+    DEBUG("[wire] %s\n", __func__);
+
+    rxBufferIndex = 0;
+    rxBufferLength = 0;
+
+    txBufferIndex = 0;
+    txBufferLength = 0;
+
+    i2c_init(ARDUINO_I2C_DEV);
+}
+
+void WirePort::begin(uint8_t addr)
+{
+    (void)addr;
+    DEBUG("[wire] %s\n", __func__);
+    core_panic(PANIC_GENERAL_ERROR, "[wire] slave mode is not supported");
+}
+
+void WirePort::setClock(uint32_t clk)
+{
+    (void)clk;
+    /* not possible, bus speed is defined by the board definition */
+    DEBUG("[wire] %s: clock is defined by board definition\n", __func__);
+}
+
+uint8_t WirePort::requestFrom(uint8_t addr, uint8_t size, uint8_t stop)
+{
+    DEBUG("[wire] %s: addr %02x, size %d, stop %d\n", __func__, addr, size, stop);
+
+    if (size > WIREPORT_BUFFER_LENGTH) {
+        size = WIREPORT_BUFFER_LENGTH;
+    }
+
+    uint8_t read = 0;
+    if (i2c_acquire(ARDUINO_I2C_DEV) == 0 &&
+        i2c_read_bytes(ARDUINO_I2C_DEV, addr, rxBuffer, size,
+                                        stop ? 0 : I2C_NOSTOP) == 0 &&
+        i2c_release(ARDUINO_I2C_DEV) == 0) {
+        read = size;
+    }
+
+    rxBufferIndex = 0;
+    rxBufferLength = read;
+
+    return read;
+}
+
+uint8_t WirePort::requestFrom(uint8_t addr, uint8_t size) {
+    return requestFrom(addr, size, 1);
+}
+
+void WirePort::beginTransmission(uint8_t addr)
+{
+    DEBUG("[wire] %s: addr %02x\n", __func__, addr);
+    transmitting = 1;
+    txAddress = addr;
+    txBufferIndex = 0;
+    txBufferLength = 0;
+}
+
+uint8_t WirePort::endTransmission(uint8_t stop)
+{
+    DEBUG("[wire] %s: stop %d\n", __func__, stop);
+
+    uint8_t wrote = 0;
+    if (i2c_acquire(ARDUINO_I2C_DEV) == 0 &&
+        i2c_write_bytes(ARDUINO_I2C_DEV, txAddress, txBuffer, txBufferLength,
```

Comment:
By checking the returned value of `i2c_write_bytes` you could know the specific error (i.e. no device responded on the address or no response after writing bytes) and return the appropriate error code.